atecc608a_se.h

```diff
@@ -34,4 +34,14 @@ psa_status_t atecc608a_init();
 
 psa_status_t atecc608a_deinit();
 
+/** Read from a given slot at an offset. Data zone has to be locked for this
+ *  function to work. */
+psa_status_t atecc608a_read(uint16_t slot, size_t offset, uint8_t *data, size_t length);
```

Comment:
Why would the driver expose a way to read/write at an offset? The API only provides a way to import/export the key as a whole.

Reply:
There is a possibility to store arbitrary data at slot 8, for example a certificate, or a signature. The user might want to write and/or read to a slot, depending on the configuration.

Reply:
Ok, but that would be exposed through the API via `psa_import_key` and `psa_export_key` with the key type `PSA_KEY_TYPE_RAW_DATA`. These functions read/write the whole slot.

Reply:
Ok, I didn't know that that's the plan. I can refactor the import_key and export_key procedures to factor for that, I would just like to get a confirmation from @Patater.

Reply:
Sounds like we don't need extra function pointers in the struct then, since we have import and export function pointers already. Let's use those.

Do we want to register a new driver struct under a different lifetime for handling raw data? or add a switch statement inside the existing driver import/export function implementation?

Reply:
There's no reason to have different lifetimes for different key types. It _might_ make sense to have different lifetimes for different slot pools, with an EC pool and a raw pool for this driver, but for the moment I think it's best to keep it simple, with a single lifetime, and hardcode which slot numbers can store which key types..

Reply:
An extra struct with different function pointers in it may end up having a smaller footprint in flash than code used to dispatch based on key type, given we already have code to dispatch based on lifetime.

Reply:
The problem with using the key export API for raw data readout is that the driver will have to export all the data that's in the slot at once, and the user might not want all of it, or the user might not even have enough space to store the data from a whole slot. Slot 8 has, for example, 416 bytes of data. If someone writes just 10 bytes, why should we export the rest?
There's a possibility to signal how much data the user wants by passing the available buffer size to the export function accordingly, but this doesn't seem to be clean for me.
What about reading only the last 10 bytes? If we want to support such thing, the export function is insufficient.

Let me know if I should refactor the export function to handle this case,

Reply:
The driver needs to remember the size of the data. The caller can ask what the size is.

What's the use case for reading only a part of the slot?

Reply:
Having a signature / arbitrary data in slot 8 that doesn't take up all of the available space. The driver has no way of knowing what is the size of such data, since it can be written during provisioning.

Reply:
If there's a “magical” provisioning mechanism that's distinct from `psa_import_key`, it needs to store the size somewhere in a way that the core or the driver can read back.

Reply:
I'd rather suggest having a separate structure / API for storing arbitrary data on a secure element, or go for a more specific version - in slots.

Reply:
Can we treat the slots as a block device block for now? This means you must write the whole block and must read the whole block. As a user, you can ignore on reads and pad as you see fit on writes.Maybe later we want to have a multi-part export for reading a subset of the data, but the API doesn't have that today.

Reply:
Ok, so that's settled. Should it be a part of the export key function dispatched in runtime, based on the key type (`PSA_KEY_TYPE_RAW_DATA`), or a separate function?

Reply:
I'd recommend separate function in any case, which you can call from the driver export function after a run-time check on the key type, or you can directly place into a separate driver struct for a separate lifetime. I'd say go with the run-time check in the driver for now, since that's easier to use from the applications perspective: they don't have to know what particular lifetime the driver decided to register the RAW export functions at.

Reply:
And what about writing to a slot? Should this also be possible only through import_key with a `PSA_KEY_TYPE_RAW_DATA` type?

Reply:
What would be a use case for writing to a slot that wouldn't go through `psa_import_key`?

Reply:
Again - writing a certificate/signature to a data slot that might be later used. In theory, we could even later use hardware accelerated operations on certificates, as 508A supports it (if the certificate is in a [compressed format](http://ww1.microchip.com/downloads/en/AppNotes/Atmel-8974-CryptoAuth-ATECC-Compressed-Certificate-Definition-ApplicationNote.pdf). It can go through import key, however it's not a key but raw data - that's why I want to be sure that we're on the same page.

Reply:
Writing a certificate would go through `import_key` with `RAW_DATA`. It's what other APIs do and I see no reason to be unusual there.

Reply:
The export_key function does not have a key type parameter. Should we infer the type in runtime by reading the hardware configuration, or should we change the API?

Reply:
The core always knows the type of a key. The application provides it when it creates a key. If the key is pre-provisioned, the provisioning mechanism provides they key type as well as the location and policy.

Reply:
I'm not sure if I understand your answer correctly. The core knows the type of the key, sure, but if it's not passed to the export_key - the function won't. How is the key type provided when the provisioning is done? If you're speaking about providing it with import_key - that's ok, but how should we store it persistently? There's no place for it in the slots. Reading the hardware configuration will only inform us about the private/public key/data properties of a slot.

Reply:
The core stores key metadata (type, policy, id, location) in internal storage.

Reply:
Again - the export_key function won't know about the key type unless it's passed to it or if it will check the hardware to determine which method of reading to use.

Reply:
Oh, you mean the `export_key` function in the driver? I guess it should receive the key attributes as an argument. Right now it only gets the size and location. Why isn't it enough to know the size and location, by the way?

Reply:
The public key is stored fragmented and padded, and has to be read using a dedicated `atcab_get_pubkey` function, that converts it to a readable format. Clear text reads are used for raw data, and exporting a private key is not even possible.

Reply:
OK, so the secure element API has to be updated. Should I do it? Will it be a part of a different PR? @Patater - are we planning on switching from https://github.com/Patater/mbed-crypto/tree/secure-element to an official branch?

Reply:
There are other pending updates to the SE API to cope with driver registration, key registration and updates to the application API. That's on me at the moment, together with https://jira.arm.com/browse/IOTCRYPT-608 and https://jira.arm.com/browse/IOTCRYPT-778